File: backend/mealroulette/schemas/taxonomy.py
from pydantic import BaseModel, Field


class FoodGroupPublic(BaseModel):
    id: str
    label: str
    description: str


class IngredientFamilyPublic(BaseModel):
    id: str
    food_group: str
    label: str
    description: str


class IngredientTaxonomySummary(BaseModel):
    id: int
    canonical_name: str
    display_name: str
    food_group: str | None
    family: str | None
    pantry_item: bool
    alias_count: int
    approved_conversion_count: int
    unapproved_conversion_count: int
    missing_family: bool
    missing_food_group: bool


class FoodGroupOverview(BaseModel):
    id: str
    label: str
    family_count: int
    ingredient_count: int
    missing_metadata_count: int


class IngredientTaxonomyOverview(BaseModel):
    totals: dict[str, int]
    food_groups: list[FoodGroupOverview]


class IngredientResolveMatch(BaseModel):
    canonical_name: str
    display_name: str
    food_group: str | None
    family: str | None
    score: float | None = None


class IngredientResolveResponseV2(BaseModel):
    status: str
    query: str
    matched_on: str | None = None
    matched_value: str | None = None
    ingredient: IngredientResolveMatch | None = None
    suggestions: list[IngredientResolveMatch] = Field(default_factory=list)


class ClassifyCandidateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=256)
    context: str | None = Field(default=None, max_length=1024)
    language: str | None = Field(default=None, max_length=16)


class ClassifySuggestion(BaseModel):
    id: str
    reason: str
    food_group: str | None = None


class ClassifyIngredientSuggestion(BaseModel):
    canonical_name: str
    display_name: str
    family: str | None
    food_group: str | None
    reason: str
    score: float | None = None


class ClassifyCandidateResponse(BaseModel):
    status: str
    query: str
    food_groups: list[ClassifySuggestion] = Field(default_factory=list)
    families: list[ClassifySuggestion] = Field(default_factory=list)
    ingredients: list[ClassifyIngredientSuggestion] = Field(default_factory=list)
    draft: dict | None = None

File: frontend/src/api/taxonomy.ts
import { apiRequest } from "./client";

export type FoodGroup = {
  id: string;
  label: string;
  description: string;
};

export type IngredientFamily = {
  id: string;
  food_group: string;
  label: string;
  description: string;
};

export type IngredientTaxonomySummary = {
  id: number;
  canonical_name: string;
  display_name: string;
  food_group: string | null;
  family: string | null;
  pantry_item: boolean;
  alias_count: number;
  approved_conversion_count: number;
  unapproved_conversion_count: number;
  missing_family: boolean;
  missing_food_group: boolean;
};

export type IngredientTaxonomyOverview = {
  totals: Record<string, number>;
  food_groups: Array<{
    id: string;
    label: string;
    family_count: number;
    ingredient_count: number;
    missing_metadata_count: number;
  }>;
};

export type IngredientResolveMatch = {
  canonical_name: string;
  display_name: string;
  food_group: string | null;
  family: string | null;
  score: number | null;
};

export type IngredientResolveV2Response = {
  status: string;
  query: string;
  matched_on?: string | null;
  matched_value?: string | null;
  ingredient?: IngredientResolveMatch | null;
  suggestions?: IngredientResolveMatch[];
};

export type ClassifyCandidateResponse = {
  status: string;
  query: string;
  food_groups: Array<{ id: string; reason: string; food_group?: string | null }>;
  families: Array<{ id: string; reason: string; food_group?: string | null }>;
  ingredients: Array<{
    canonical_name: string;
    display_name: string;
    family: string | null;
    food_group: string | null;
    reason: string;
    score: number | null;
  }>;
  draft: Record<string, unknown> | null;
};

export function fetchFoodGroups(token: string): Promise<FoodGroup[]> {
  return apiRequest<FoodGroup[]>("/api/food-groups", { token });
}

export function fetchFoodGroupFamilies(token: string, foodGroupId: string): Promise<IngredientFamily[]> {
  return apiRequest<IngredientFamily[]>(`/api/food-groups/${encodeURIComponent(foodGroupId)}/families`, { token });
}

export function fetchFamilyIngredients(token: string, familyId: string): Promise<IngredientTaxonomySummary[]> {
  return apiRequest<IngredientTaxonomySummary[]>(
    `/api/ingredient-families/${encodeURIComponent(familyId)}/ingredients`,
    { token },
  );
}

export function fetchIngredientTaxonomyOverview(token: string): Promise<IngredientTaxonomyOverview> {
  return apiRequest<IngredientTaxonomyOverview>("/api/ingredient-taxonomy/overview", { token });
}

export function resolveIngredientV2(token: string, proposedName: string): Promise<IngredientResolveV2Response> {
  return apiRequest<IngredientResolveV2Response>("/api/ingredients/resolve-v2", {
    method: "POST",
    token,
    body: JSON.stringify({ proposed_name: proposedName }),
  });
}

export function classifyIngredientCandidate(
  token: string,
  payload: { name: string; context?: string; language?: string },
): Promise<ClassifyCandidateResponse> {
  return apiRequest<ClassifyCandidateResponse>("/api/ingredients/classify-candidate", {
    method: "POST",
    token,
    body: JSON.stringify(payload),
  });
}
